src/median.ts

```diff
@@ -0,0 +1,86 @@
+import type {
+  IterableContainer,
+  NonEmptyArray,
+  NonEmptyReadonlyArray,
+} from "./internal/types";
+import { purry } from "./purry";
+
+type Median<T extends IterableContainer<number>> = T extends
+  | NonEmptyArray<number>
+  | NonEmptyReadonlyArray<number>
+  ? number
+  : T extends [] | readonly []
+    ? undefined
+    : number | undefined;
```

Comment:
I'm not a big fan of the NonEmptyArray usage because it obfuscates the logic behind another reference hop, making local modifications harder when needed. The code is even shorter without it. Additionally, instead of repeating the different components (undefined, number) several times, we can be more concise by conditionally adding each one exactly one.

```suggestion
type Median<T extends IterableContainer<number>> =
  | (T extends readonly [] ? never : number)
  | (T extends readonly [unknown, ...Array<unknown>] ? never : undefined);
```

This reads: "If the array is empty then the result can never be number, and if the array is non-empty the result can never be undefined.

Reply:
That seems better, thanks! I used these types initially to cover all the cases explicitly, but I can see how your approach simplifies the logic.